Construct and add 'LABEL' command from 'label' option to last stage

arm64b opened the pull request:
fixes #36996
fixes https://github.com/docker/for-linux/issues/305
    
Currently for multi-stage build, if `--target` specified, the `--label` option will be ignored. The root cause is the last stage build will remove the `LABEL` command(s) node created from the `--label` option. In order to address this issue, we can re-add the `LABEL` command(s) to the last stage.

Also, add a new test case to cover the fix.
    
Signed-off-by: Dennis Chen <dennis.chen@arm.com>
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Resolve the issue #36996  
**- How I did it**
Create new `LABEL` command(s) and add it to the last stage.
**- How to verify it**
integration test
**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

arm64b committed:
Add 'LABEL' command from '--label' to the last stage

This PR is tring to fix issue #36996.

Currently for multi-stage build, if `--target` specified, the `--label` option
will be ignored. The root cause is the last stage build will remove the `LABEL`
command(s) node created from the `--label` option. In order to address this issue,
we can create `LABEL` command(s) and add it/tem to the last stage.

Signed-off-by: Dennis Chen <dennis.chen@arm.com>

arm64b committed:
Add test case for `--label` with `--target`

Add a new test case `TestBuildLabelWithTargets` to cover the Docker builder
with both `--label` and `--target` options.

Signed-off-by: Dennis Chen <dennis.chen@arm.com>

arm64b commented:
> 08:10:06 The following new tests were added to integration-cli:
08:10:06 
08:10:06 +func (s *DockerSuite) TestBuildLabelWithTargets(c *check.C) {
08:10:06 
08:10:06 integration-cli is deprecated. Please add an API integration test to
08:10:06 ./integration/COMPONENT/. See ./TESTING.md for more details.
08:10:06 
08:10:06 Build step 'Execute shell' marked build as failure

Hello @vdemeester , How can I adapt the new test case to the existing integration test to avoid above error? According to the result of other nodes, seems it's not related with the code, I guess it should be depends on the testing script ❓

ijc commented on builder/dockerfile/instructions/commands.go:
```diff
@@ -119,6 +119,21 @@ type LabelCommand struct {
 }
 
+// NewLableCommand creates a new 'LABLE' command
+func NewLableCommand(k string, v string) *LabelCommand {
```
You have "Lable" rather than "Label" three times in these two lines.

arm64b replied:
Hello @ijc , thanks for spotting this  😛 will correct in the 2nd version...

tonistiigi commented on builder/dockerfile/builder.go:
```diff
@@ -229,6 +229,12 @@ func (b *Builder) build(source builder.Source, dockerfile *parser.Result) (*buil
 			return nil, errdefs.InvalidParameter(errors.Errorf("failed to reach build target %s in Dockerfile", b.options.Target))
 		}
 		stages = stages[:targetIx+1]
+		// Add `LABEL` command if `--label` option specified
```
We should use this for the case when `--target` is not set as well and remove `addNodesForLabelOption`.

arm64b replied:
OK, I will remove the `addNodesForLabelOption` and use `NewLabelCommand` to add the `--label` options into the last stage, thus we will have the `LABEL` there.

arm64b committed:
Remove unused 'label' related functions

Since we use `NewLabelCommand()` instead of `addNodesForLabelOption()`
to create the 'LABEL' commands from '--label' options, so all the related
functions should be removed.

Signed-off-by: Dennis Chen <dennis.chen@arm.com>

tonistiigi commented:
@arm64b 

> How can I adapt the new test case to the existing integration test to avoid above error? According to the result of other nodes, seems it's not related with the code, I guess it should be depends on the testing script ❓

Add new tests to the `integration` pkg, not the `integration-cli`. See https://github.com/moby/moby/blob/17b039cb4901fc0a37c409ebfdf4f3b2fcec4409/integration/build/build_test.go for examples.

tonistiigi commented on builder/dockerfile/builder.go:
```diff
@@ -208,13 +209,30 @@ func newBuilder(clientCtx context.Context, options builderOptions) *Builder {
 
+// Build 'LABEL' command(s) from '--label' options and add to the last stage
+func buildLabelOptions(labels map[string]string, stages []instructions.Stage) {
+	if len(labels) > 0 {
```
nit: this `if` is unnecessary

tonistiigi commented on builder/dockerfile/builder.go:
```diff
@@ -208,13 +209,30 @@ func newBuilder(clientCtx context.Context, options builderOptions) *Builder {
+		sort.Strings(keys)
+		for _, key := range keys {
+			value := labels[key]
+			// Value must be single quoted to prevent env variable expansion
```
Instead of this quoting hack, could we add `NewLabelCommand(key, value string, expand bool)` so that when `expand` is false it would make `label.Expand()` no-op.

arm64b replied:
Thanks very much for the review! 👍  Will update the change per the comment...

tonistiigi commented on builder/dockerfile/instructions/commands.go:
```diff
@@ -110,6 +110,25 @@ type MaintainerCommand struct {
+// NewLabelCommand creates a new 'LABEL' command
+func NewLabelCommand(k string, v string, expand bool) *LabelCommand {
+	// Value must be single quoted to prevent env variable expansion if not expected
+	if !expand {
```
What I meant for this parameter was to set it as a property of `LabelCommand`, like `& LabelCommand{expand: expand, ...}` and then change the expand command to:

```
 func (c *LabelCommand) Expand(expander SingleWordExpander) error {
        if !c.expand {
           return nil
        }
 	return expandKvpsInPlace(c.Labels, expander)
 }
```

arm64b replied:
Ah, my fault to misunderstand that,  I'll add the `expand` as property to the `LabelCommand` and leave the expand decision to the `dispatch` component according to the property value. Nice 😄

tonistiigi commented:
LGTM

cpuguy83 commented:
Test failures look related.

arm64b commented:
Out of office, will take a look after back next week

arm64b commented:
`TestBuildHistory` failure: 
> 23:36:44 FAIL: docker_cli_history_test.go:16: DockerSuite.TestBuildHistory
23:36:44 
23:36:44 docker_cli_history_test.go:53:
23:36:44     c.Assert(actualValue, checker.Contains, echoValue)
23:36:44 ... obtained string = "66d2fb3987f5        1 second ago        /bin/sh -c #(nop)  LABEL label.Z=\"Z\"   0B                  "
23:36:44 ... substring string = "LABEL label.Z=Z"

The failure reason comes from different behavior of the `LABEL` command before and after the PR --
Before the PR:
`LABEL label.Z="Z"` will generate "LABEL label.Z=Z"
While after this PR:
`LABEL label.Z="Z"` will generate "LABEL label.Z=\"Z\""

The same as `TestImagesFilterLabelMatch` case.

@tonistiigi what's your suggestion about this? IMO, we can remove the new added `expand` property and use the original single quote to prevent env from expanding...

tonistiigi commented:
@arm64b Does it come from the `parseLabel()` not setting `expand` properly? It may make sense to reverse the bool to `noExpand` to avoid accidental behavior on zero value. It would be fine to change the test if the quotes changed for `--label` as the formatting for that is undefined anyway but that test doesn't seem to use `--label`.

arm64b commented:
Hello @tonistiigi , just saw your comments. I've figure out another solution to align the behavior of inline `LABEL` command with the one from `--label` option in `parseLabel()`, with single quoted value and set the `expand` to true, then leave the `dispatch` component to handle the case failed.
don't know if does make sense or not...

arm64b commented:
@tonistiigi any suggestion for the latest change? I will update if any

tonistiigi commented on builder/dockerfile/instructions/parse.go:
```diff
@@ -228,9 +228,31 @@ func parseLabel(req parseRequest) (*LabelCommand, error) {
 		return nil, err
 	}
 
+	kvpMap := make(map[string]string)
```
What is the reason for the lines 231-251 here? Why wouldn't just setting `expand` to true work?

arm64b replied:
The reason for the lines 231-251 is to make sure the env variable LABEL like this "LABEL foo="$PATH" don't expand, so we can get the expected "$PATH" while not something like "/usr/bin:/usr/local/bin" because `expand=true` will trigger the expand handling method if no single quoted way by lines 231-251, which is different from the `--label` option where the `expand=false` meaning the env variable will not be expanded.

tonistiigi replied:
But the variables are expanded to `LABEL` values with the current implementation. If you think this should be changed it doesn't belong in this PR.

arm64b replied:
>But the variables are expanded to LABEL values with the current implementation

Don't quite understand here. If the `current implementation` is the PR's, then the env variables will not be expanded, this is the same as the original code before this PR. Correct me if I am wrong here.

>If you think this should be changed it doesn't belong in this PR.

If we don't have this change in this PR, then we will have 2 label related test cases failure as we saw before.
But if you have some more reasonable solution, I am happy to implement that.

tonistiigi replied:
```
/tmp # cat Dockerfile
from busybox
label foo="$PATH"
/tmp # docker inspect -f '{{.Config.Labels}}' $(docker build -q .)
map[foo:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin]
/tmp # docker version -f '{{ .Server.Version}}'
18.03.1-ce
```

arm64b replied:
Ah, that's because the client side will expand the env variable automatically  and `POST` the expanded `LABEL` value to the server, if we apply this PR, the result is the same as `18.03.1-ce`

tonistiigi replied:
No, we definitely do not evaluate any variables inside a Dockerfile in client side.

arm64b replied:
What I said is the client  side will expand the env variable of `LABEL` command then `POST` it to the server, look like this: 
```
Calling POST /v1.35/build?buildargs=%7B%7D&cachefrom=%5B%5D&cgroupparent=&cpuperiod=0&cpuquota=0&cpusetcpus=&cpusetmems=&cpushares=0&dockerfile=Dockerfile&labels=%7Buser%7Dbin%7D...
```

tonistiigi replied:
`labels=` in the POST query are the extra labels from `--label` flag(not expanded by client as well, could be by the shell when using cli but that is irrelevant) not the values from `LABEL` command. Even with this PR these never call this `parseLabel` function that is being discussed here.

arm64b replied:
Hmm, I just checked the debug message of the daemon, seems the `--label foo="$PATH"` will expand to `labels=%7B%22bbc%22%3A%22%2Fgo%2Fbin%3A%2Fusr%2Flocal%2Fsbin...` when `POST`, the `LABEL` inside the Dockerfile will not. Sorry for the misleading info 😄

arm64b replied:
But the `LABEL foo="$PATH"` inside the Dockerfile has the same behavior before and after this PR

arm64b replied:
So @tonistiigi , now we can focus on the `parseLabel` function, if we just set `expand=true`, we will fail the CI as you know 😭

tonistiigi replied:
What test will fail and why?

arm64b commented:
@tonistiigi From the  CI test result, seems we can simply set the `expand=true` in `parseLabel()` command.

tonistiigi commented on builder/dockerfile/instructions/commands.go:
```diff
@@ -110,17 +110,37 @@ type MaintainerCommand struct {
 }
 
+// NewLabelCommand creates a new 'LABEL' command
+func NewLabelCommand(k string, v string, exp bool) *LabelCommand {
```
nit: rename `exp` variable here as well

tonistiigi reviewed (approved):
LGTM

thaJeztah reviewed (approved):
LGTM after the nit is addressed 🤗

thaJeztah commented:
PowerPC https://jenkins.dockerproject.org/job/Docker-PRs-powerpc/9927/console failing on  flaky (https://github.com/moby/moby/issues/32673)

```
20:27:11 FAIL: docker_api_swarm_test.go:296: DockerSwarmSuite.TestAPISwarmLeaderElection
20:27:11 
20:27:11 [dded101a9941d] waiting for daemon to start
20:27:11 [dded101a9941d] daemon started
20:27:11 
20:27:11 [dd7cb1651984a] waiting for daemon to start
20:27:11 [dd7cb1651984a] daemon started
20:27:11 
20:27:11 [d908d5be69f79] waiting for daemon to start
20:27:11 [d908d5be69f79] daemon started
20:27:11 
20:27:11 [dded101a9941d] exiting daemon
20:27:11 assertion failed: error is not nil: Error response from daemon: rpc error: code = DeadlineExceeded desc = context deadline exceeded
20:27:11 [dd7cb1651984a] exiting daemon
20:27:11 [d908d5be69f79] exiting daemon
20:27:32
```

jeffrey-aguilera commented:
This is still broken going into 2023.
